jwst/background/background_sub.py

```diff
@@ -26,85 +26,171 @@ def background_sub(input_model, bkg_list, sigma, maxiters):
+        For input rateint files: 3,
+
     Returns:
     --------
-
     avg_bkg: data model
         The averaged background exposure
-
     """
-
     num_bkg = len(bkg_list)
     avg_bkg = None
-    cdata = None
 
-    # Loop over the images to be used as background
-    for i, bkg_file in enumerate(bkg_list):
-        log.info(f'Accumulate bkg from {bkg_file}')
-        bkg_model = datamodels.ImageModel(bkg_file)
+    if data_dim == 2:  # background rate input
+        cdata = None
+        # Loop over the images to be used as background
+        for i, bkg_file in enumerate(bkg_list):
+            log.info(f'Accumulate bkg from {bkg_file}')
+            bkg_model = datamodels.ImageModel(bkg_file)
+
+            # Initialize the avg_bkg model, if necessary
+            if avg_bkg is None:
+                avg_bkg = datamodels.ImageModel(bkg_model.shape)
+
+            if cdata is None:
+                cdata = np.zeros(((num_bkg,) + bkg_model.shape))
+                cerr = cdata.copy()
 
-        # Initialize the avg_bkg model, if necessary
-        if avg_bkg is None:
-            avg_bkg = datamodels.ImageModel(bkg_model.shape)
+            # Accumulate the data from this background image
+            cdata[i] = bkg_model.data
+            cerr[i] = bkg_model.err * bkg_model.err
 
-        if cdata is None:
-            cdata = np.zeros(((num_bkg,) + bkg_model.shape))
-            cerr = cdata.copy()
+            avg_bkg.dq = np.bitwise_or(avg_bkg.dq, bkg_model.dq)
 
-        # Accumulate the data from this background image
-        cdata[i] = bkg_model.data
-        cerr[i] = bkg_model.err * bkg_model.err
-        avg_bkg.dq = np.bitwise_or(avg_bkg.dq, bkg_model.dq)
+            bkg_model.close()
 
-        bkg_model.close()
+        # Clip the background data
+        log.debug('clip with sigma={} maxiters={}'.format(sigma, maxiters))
+        mdata = sigma_clip(cdata, sigma=sigma, maxiters=maxiters, axis=0)
 
-    # Clip the background data
-    log.debug('clip with sigma={} maxiters={}'.format(sigma, maxiters))
-    mdata = sigma_clip(cdata, sigma=sigma, maxiters=maxiters, axis=0)
+        # Compute the mean of the non-clipped values
+        avg_bkg.data = mdata.mean(axis=0).data
 
-    # Compute the mean of the non-clipped values
-    avg_bkg.data = mdata.mean(axis=0).data
+        # Mask the ERR values using the data mask
+        merr = np.ma.masked_array(cerr, mask=mdata.mask)
 
-    # Mask the ERR values using the data mask
-    merr = np.ma.masked_array(cerr, mask=mdata.mask)
+        # Compute the combined ERR as the uncertainty in the mean
+        avg_bkg.err = (np.sqrt(merr.sum(axis=0)) / (num_bkg - merr.mask.sum(axis=0))).data
 
-    # Compute the combined ERR as the uncertainty in the mean
-    avg_bkg.err = (np.sqrt(merr.sum(axis=0)) / (num_bkg - merr.mask.sum(axis=0))).data
+        return avg_bkg
 
-    return avg_bkg
+    else:  # data_dim = 3; background rateints input
```

Comment:
I think this could be structured in a much more simple way. You don't really need to completely separate branches to do all the steps depending on whether the bkg models are 2D or 3D. You could keep the original loop over the individual bkg files and then simply insert a check here https://github.com/spacetelescope/jwst/blob/master/jwst/background/background_sub.py#L82 that checks the dimensions of `bkg_model.data` and if it's 3D, then insert the new code that first averages the integrations of that bkg file into a single 2D image and then store the result in the accumulators `cdata` and `cerr`, just as it's done if the bkg_model is 2D. And then do the remaining sigma-clipping and averaging on the stack of 2D bkg images as before (i.e. using the existing code starting at https://github.com/spacetelescope/jwst/blob/master/jwst/background/background_sub.py#L90

Reply:
Right. Code has been simplified.